### PR TITLE
Refactor buildtool config

### PR DESCRIPTION
**Pull request by Nightenom**
# Changes proposed in this pull request:
- Move current buildtool settings to forge config, add more settings into buildtool
- Enhance forge config (can/should be copied to mcol):
  - add watcher logic for config values
  - borrow/adapt/overcome default comments from mcol
  - do not create client config on dedi server so client config can have client classes
- introduce indexed component via MessageFormat from ICU which has support for plurals etc (should be safe to use in vanilla, is meant to be also used in vanilla, but for full support we would need to mixin for json serializer, also not sure about network serialization like commands)
- remove few 100% dead translations

Did not test multiplayer (for preview sharing) but it's 99.9% refactor without intentional logic changes

Review please

### CODE REVIEW

File: src/main/java/com/ldtteam/structurize/client/gui/AbstractBlueprintManipulationWindow.java [hunk 1]
```diff
@@ -314,46 +322,118 @@ protected void settingsClicked()
...
+                buttonImage.setText(Component.translatable(value ? "options.on" : "options.off"));
                 buttonImage.setHandler((button) -> {
-                    settings.get(index).setValue(!settings.get(index).getValue());
-                    Network.getNetwork().sendToServer(new SyncSettingsToServer());
-                    RenderingCache.getOrCreateBlueprintPreviewData(bluePrintId).syncChangesToServer();
+                    final Boolean newValue = !typedSetting.getValue(blueprintRenderSettings);
+                    typedSetting.setValue(blueprintRenderSettings, newValue);
+                    buttonImage.setText(Component.translatable(newValue ? "options.on" : "options.off"));
+
+                    // TODO: ray why? RenderingCache.getOrCreateBlueprintPreviewData(bluePrintId).syncChangesToServer();
```
Comment (Nightenom) [inline src/main/java/com/ldtteam/structurize/client/gui/AbstractBlueprintManipulationWindow.java]: @Raycoms

Comment (Raycoms) [reply src/main/java/com/ldtteam/structurize/client/gui/AbstractBlueprintManipulationWindow.java]: The moment you turn this on, its not instantly shared if you dont do this. It would then only be shared on a bigger blueprint change

File: src/main/java/com/ldtteam/structurize/client/gui/WindowExtendedBuildTool.java [hunk 2]
```diff
@@ -466,6 +466,20 @@ public void settingsClicked()
         folderList.hide();
         blueprintList.disable();
         blueprintList.hide();
+
+        final View cats = findPaneOfTypeByID("categories", View.class);
```
Comment (Raycoms) [inline src/main/java/com/ldtteam/structurize/client/gui/WindowExtendedBuildTool.java]: What does this do?

File: src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java [hunk 3]
```diff
@@ -34,10 +54,11 @@ private BlueprintRenderSettings()
      */
     public BlueprintRenderSettings(final FriendlyByteBuf byteBuf)
     {
-        final int size = byteBuf.readInt();
-        for (int i = 0; i < size; i++)
+        renderSettings = new ArrayList<>(SharedSettings.VALUES.size());
+
+        for (final var setting : SharedSettings.VALUES)
```
Comment (Raycoms) [inline src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java]: I really hate var, lets try to use the instance if we actually know it =D

File: src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java [hunk 4]
```diff
@@ -47,21 +68,147 @@ public BlueprintRenderSettings(final FriendlyByteBuf byteBuf)
...
+    public <T> T getSetting(final SharedSettings<T> key)
+    {
+        return renderSettings == null ?
+            // client instance
+            key.forgeConfig.get() :
+            // server instance (or client instance from buffer)
+            (T) renderSettings.get(key.id);
+    }
+
+    public <T> void setSetting(final SharedSettings<T> key, final T value)
```
Comment (Raycoms) [inline src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java]: Docs on these methods.

File: src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java [hunk 5]
```diff
@@ -47,21 +68,147 @@ public BlueprintRenderSettings(final FriendlyByteBuf byteBuf)
...
     }
 
     /**
-     * Register a new setting.
-     * @param key the settings key.
-     * @param value the settings value.
+     * Wrapper to merge normal and shared configs
      */
-    public void registerSetting(final String key, final boolean value)
+    public record EitherConfig<T>(ConfigValue<T> normal, SharedSettings<T> wrapped)
```
Comment (Raycoms) [inline src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java]: Im not sure I get what all of this boilerplate here is for?

File: src/main/java/com/ldtteam/structurize/util/IndexedTranslatableContents.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+import java.text.MessageFormat;
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Inspired by {@link TranslatableContents} to support things from {@link MessageFormat}
+ */
+public class IndexedTranslatableContents implements ComponentContents
+{
```
Comment (Raycoms) [inline src/main/java/com/ldtteam/structurize/util/IndexedTranslatableContents.java]: What is this for?

File: src/main/java/com/ldtteam/structurize/client/gui/AbstractBlueprintManipulationWindow.java (hunk 1 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/client/gui/AbstractBlueprintManipulationWindow.java]: Oh,kk, so it's basically config watcher thing, will move

File: src/main/java/com/ldtteam/structurize/client/gui/WindowExtendedBuildTool.java (hunk 2 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/client/gui/WindowExtendedBuildTool.java]: Hides categories when settings are displayed to avoid visual clutter

File: src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java (hunk 3 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java]: Yep, I'm like foreach are fine, but will switch

File: src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java (hunk 5 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java]: You made it, I just kept functionality, u tell me :P
Like for now it's just one boolean, but this allows any thing to be sent through

File: src/main/java/com/ldtteam/structurize/util/IndexedTranslatableContents.java (hunk 6 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/util/IndexedTranslatableContents.java]: Look en_us.json for default keys, it's the classic {index, format, etc} translating instead of dumb %letter thing

File: src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java (hunk 5 above)
Comment (Raycoms) [reply src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java]: I mean the ton of new settings and records and stuff here =D

File: src/main/java/com/ldtteam/structurize/util/IndexedTranslatableContents.java (hunk 6 above)
Comment (Raycoms) [reply src/main/java/com/ldtteam/structurize/util/IndexedTranslatableContents.java]: And we need all of that just to pick up a translation ?

Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/util/IndexedTranslatableContents.java]: Yeah, but it's ten times better cuz % thing is far from decent for more than half of languages

Comment (Raycoms) [reply src/main/java/com/ldtteam/structurize/util/IndexedTranslatableContents.java]: Yeah, but I think its not worth to stray from the standard that we are using everywhere

File: src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java (hunk 5 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java]: to keep ur functionality and merge (for gui) normal configvalues with synced/shared configs, i was also thinking about reducing to just one boolean (cuz it was one boolean in link-sessions anyway), but u made it like this so i kept it logic-wise same, just binary incompatible cuz normal configvalues

File: src/main/java/com/ldtteam/structurize/util/IndexedTranslatableContents.java (hunk 6 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/util/IndexedTranslatableContents.java]: we definitely should ditch percent if we can, message format from icu introduces proper plural etc. see <https://support.crowdin.com/icu-message-syntax/>

File: src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java (hunk 5 above)
Comment (Raycoms) [reply src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java]: Do we use those in minecolonies? I mean, I would make this now just "only client configs" and get rid of the old one.

File: src/main/java/com/ldtteam/structurize/config/AbstractConfiguration.java [hunk 7]
```diff
@@ -91,11 +107,76 @@ protected static <T> ConfigValue<List<? extends T>> defineList(
...
-        return buildBase(builder, key).defineEnum(key, defaultValue);
+        return buildBase(builder,
+            key,
+            IcuComponent.of(DEFAULT_KEY_PREFIX + "enum",
+                defaultValue,
+                Arrays.stream(defaultValue.getDeclaringClass().getEnumConstants()).map(Enum::name).collect(Collectors.joining(", "))))
+            .defineEnum(key, defaultValue);
+    }
+
+    protected <T> void addWatcher(final ConfigValue<T> configValue, final ConfigListener<T> listener)
```
Comment (someaddons) [inline src/main/java/com/ldtteam/structurize/config/AbstractConfiguration.java:126]: Why do we have this odd watcher addition? Forge configs already have reloading on change, and an event that fires when it does. Also includes value caching

Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/config/AbstractConfiguration.java:126]: yeah, but the event is fired for everything, you have no way to attach calls for static instances, eg. <https://github.com/ldtteam/Structurize/pull/613/files#diff-32d1c580005cc8075e20d9ae1ba1eff01b329e665837933039b6c76589c12224R42>
or i didnt find a way to do it

edit: plus the mentioned event drives this watcher logic

File: src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java (hunk 5 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/config/BlueprintRenderSettings.java]: Entire class deleted based on discord convo

File: src/main/java/com/ldtteam/structurize/config/AbstractConfiguration.java (hunk 7 above)
Comment (someaddons) [reply src/main/java/com/ldtteam/structurize/config/AbstractConfiguration.java:126]: not sure I understand why you need to attach a watcher instead of using ModConfigEvent.Reloading?

Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/config/AbstractConfiguration.java:126]: repeating myself:
> edit: plus the mentioned event drives this watcher logic

this watcher is not a file watcher, its just class for checking if the value returned by forge config has changed, the event has no info about what was changed or what not, the only thing it has is the file that got changed, not which values or even if they actually have changed

Comment (someaddons) [reply src/main/java/com/ldtteam/structurize/config/AbstractConfiguration.java:126]: why do you need to know if a specific value changed? just reload everything

Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/config/AbstractConfiguration.java:126]: well, im not gonna reconstruct our entire blueprint gpu cache just because irrelevant config has changed, if it would 
be off-thread then mby fine, but renderer is not there yet..

edit: so im not gonna send entire client previews to server just because random renderer setting has changed

File: src/main/java/com/ldtteam/structurize/client/gui/AbstractBlueprintManipulationWindow.java [hunk 8]
```diff
@@ -314,46 +322,117 @@ protected void settingsClicked()
...
+                    }
+
+                    if (testResult)
+                    {
+                        // need properly typed thing now, but it's validated so parsing should never crash
+                        // TODO: switch instanceof
+                        final Object oldValue = typedSetting.get();
+                        if (oldValue instanceof Integer)
+                        {
+                            Structurize.getConfig().set(typedSetting, newValue.intValue());
```
Comment (Raycoms) [inline src/main/java/com/ldtteam/structurize/client/gui/AbstractBlueprintManipulationWindow.java:420]: Feel like this util should be in the config and not here.

File: src/main/java/com/ldtteam/structurize/storage/ClientStructurePackLoader.java [hunk 9]
```diff
@@ -340,7 +340,7 @@ public static Path zipSlipProtect(ZipEntry zipEntry, Path targetDir) throws IOEx
     public static void handleSaveScanMessage(final CompoundTag compound, final String fileName)
     {
         final String packName = Minecraft.getInstance().getUser().getName().toLowerCase(Locale.US);
-        StructurePacks.selectedPack = StructurePacks.getStructurePack(Minecraft.getInstance().getUser().getName());
+        
         RenderingCache.getOrCreateBlueprintPreviewData("blueprint").setBlueprintFuture(
```
Comment (Raycoms) [inline src/main/java/com/ldtteam/structurize/storage/ClientStructurePackLoader.java]: Shouldn't we have this here?

File: src/main/java/com/ldtteam/structurize/storage/rendering/RenderingCache.java [hunk 10]
```diff
@@ -122,15 +125,15 @@ public static BlueprintPreviewData getOrCreateBlueprintPreviewData(final String
...
+        return Structurize.getConfig().getClient().rendererLightLevel.get() >= 0;
     }
 
     /**
      * @return static light level
      */
     public static int getOurLightLevel()
     {
-        return 15;
+        return Structurize.getConfig().getClient().rendererLightLevel.get();
```
Comment (Raycoms) [inline src/main/java/com/ldtteam/structurize/storage/rendering/RenderingCache.java:136]: I think having this in the client config is fine, Im not sure I would have it in the build tool config.

File: src/main/java/com/ldtteam/structurize/util/IcuTranslatableContents.java [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+import java.util.Arrays;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ * Inspired by {@link TranslatableContents} to support things from {@link MessageFormat} (plurals, pronouns, etc.)
+ */
+public class IcuTranslatableContents implements ComponentContents
```
Comment (Raycoms) [inline src/main/java/com/ldtteam/structurize/util/IcuTranslatableContents.java]: Would probably put that in a separate PR

File: src/main/java/com/ldtteam/structurize/storage/ClientStructurePackLoader.java (hunk 9 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/storage/ClientStructurePackLoader.java]: oh it was setter im just blind

File: src/main/java/com/ldtteam/structurize/storage/rendering/RenderingCache.java (hunk 10 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/storage/rendering/RenderingCache.java:136]: it's heavily related tho

Comment (Raycoms) [reply src/main/java/com/ldtteam/structurize/storage/rendering/RenderingCache.java:136]: Yeah, the thing is that I would call this more of a "special" setting, rather than sth that the avg user would touch

Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/storage/rendering/RenderingCache.java:136]: can add nested settings and move also worldgen/fluid placeholder there, tho i would leave air placeholders in the main one cuz it has more impact

File: src/main/java/com/ldtteam/structurize/util/IcuTranslatableContents.java (hunk 11 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/util/IcuTranslatableContents.java]: can do later, but meh

File: src/main/java/com/ldtteam/structurize/storage/rendering/RenderingCache.java (hunk 10 above)
Comment (Raycoms) [reply src/main/java/com/ldtteam/structurize/storage/rendering/RenderingCache.java:136]: So, I think the worldgen/fluid ones is nice to have, but maybe we could have a common setting for that?

Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/storage/rendering/RenderingCache.java:136]: u mean merge air/solid/fluid placeholders into "render placeholders nice" which would hide air/render fluid/render worldgen?

Comment (Raycoms) [reply src/main/java/com/ldtteam/structurize/storage/rendering/RenderingCache.java:136]: Ye

File: src/main/java/com/ldtteam/structurize/util/IndexedTranslatableContents.java (hunk 6 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/util/IndexedTranslatableContents.java]: Will PR alone

File: src/main/java/com/ldtteam/structurize/storage/rendering/RenderingCache.java (hunk 10 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/storage/rendering/RenderingCache.java:136]: Merged into one, there are only four now, which looks fine from ui perspective (better than six :smile:)

File: src/main/java/com/ldtteam/structurize/util/IcuTranslatableContents.java (hunk 11 above)
Comment (Nightenom) [reply src/main/java/com/ldtteam/structurize/util/IcuTranslatableContents.java]: Done

**Review by Raycoms — APPROVED** [review]

**Review by someaddons — APPROVED** [review]
